borsh-derive-internal/src/enum_de.rs

```diff
@@ -82,32 +112,59 @@ pub fn enum_de(input: &ItemEnum, cratename: Ident) -> syn::Result<TokenStream2>
     } else {
         quote! {}
     };
+    if use_discriminant {
+        Ok(quote! {
+            impl #impl_generics #cratename::de::BorshDeserialize for #name #ty_generics #where_clause {
+                fn deserialize_reader<R: borsh::__private::maybestd::io::Read>(reader: &mut R) -> ::core::result::Result<Self, #cratename::__private::maybestd::io::Error> {
+                    let tag = <u8 as #cratename::de::BorshDeserialize>::deserialize_reader(reader)?;
+                    <Self as #cratename::de::EnumExt>::deserialize_variant(reader, tag)
+                }
+            }
 
-    Ok(quote! {
-        impl #impl_generics #cratename::de::BorshDeserialize for #name #ty_generics #where_clause {
-            fn deserialize_reader<R: borsh::__private::maybestd::io::Read>(reader: &mut R) -> ::core::result::Result<Self, #cratename::__private::maybestd::io::Error> {
-                let tag = <u8 as #cratename::de::BorshDeserialize>::deserialize_reader(reader)?;
-                <Self as #cratename::de::EnumExt>::deserialize_variant(reader, tag)
+            impl #impl_generics #cratename::de::EnumExt for #name #ty_generics #where_clause {
+                fn deserialize_variant<R: #cratename::__private::maybestd::io::Read>(
+                    reader: &mut R,
+                    variant_tag: u8,
+                ) -> ::core::result::Result<Self, #cratename::__private::maybestd::io::Error> {
+                    let mut return_value =
+                        #variant_arms {
+                        return Err(#cratename::__private::maybestd::io::Error::new(
+                            #cratename::__private::maybestd::io::ErrorKind::InvalidInput,
+                            #cratename::__private::maybestd::format!("Unexpected variant tag: {:?}", variant_tag),
+                        ))
+                    };
+                    #init
+                    Ok(return_value)
+                }
+            }
+        })
```

Comment:
There is seem to be 90% duplicated code in this and the next branch. Please, avoid code duplication as it is easy to introduce a bug this way in the future.